### ISSUE TITLE
VCS 3.4.17 Hangs During Build (Access Bug)

### ISSUE THREAD

**Issue by alanv73** [joyfullservice/msaccess-vcs-integration#302]
VCS Hangs during build process. I feel sure it's an object in my repo that's hanging it, but I can't tell what. If I had access to a build log I might be able to narrow it down.

![image](https://user-images.githubusercontent.com/29257817/152652103-affe36f7-41a3-48dc-97bc-0a8ec4b13b95.png)

Edit: Deleted a bunch of the tabledef files related to the tables I've been working with recently and now it hangs at a different place.

![image](https://user-images.githubusercontent.com/29257817/152652507-c7c888be-f515-4dc2-a9a9-64214ead6ed4.png)

**Comment by hecon5** [comment]
I've had this start happening to me, too. [However, I'm 99% sure it's due to a recent bug:](https://www.devhut.net/another-access-bug-cannot-open-any-more-databases/) 
Are you on M365?

**Comment by alanv73** [comment]
OS is Win 10. Office 365.

**Comment by hecon5** [comment]
Yeah, pretty sure that's it. Mine started when I got updated to the version in question. If you have the same one, I bet it's that...
Leave this open until they patch it (hopefully soon!), so we can keep an eye on it.

**Comment by joyfullservice** [comment]
For anyone out there that is dealing with the [Access bug](https://www.devhut.net/another-access-bug-cannot-open-any-more-databases/), I wanted to point out how this add-in handles trusted locations. By default, the install folder for this add-in is added as a trusted location when the add-in is installed for the first time. (Note that this is a dedicated folder for this add-in, not a folder shared with other add-ins.) The source code for managing the trusted location can be found [here](https://github.com/joyfullservice/msaccess-vcs-integration/blob/16c156db50e30a3efb328b233eacba1a2918c7b8/Version%20Control.accda.src/modules/modInstall.bas#L574).

![image](https://user-images.githubusercontent.com/7254992/152838910-262ae787-cb32-446b-b29c-dafa90b78abe.png)

Based on what I am seeing in discussions online, you will probably also need to trust the location of the database you are exporting or building. Perhaps someone affected by this issue could do some testing and comment on whether this is able to work around the problem.

**Comment by alanv73** [comment]
I'm not sure how this became an Access bug. I was able to make it do a complete build by deleting a few of the table links from the database and re-create them (lucky they were just links), then re-export the database and build from that. It doesn't address the problem that VCS is not reporting a problem with my database (or the repo it's building from) during the build process, it just freezes and never completes. If there's a problem, tell me there's a problem and error out. It would also be helpful if there was a build log that I could view to see where in the process it was having the problem. As it is I have to systematically remove items from the database one by one until it no longer hangs.

**Comment by hecon5** [comment]
It's because (behind the scenes) the VCS uses objects and connections (ADODB) to build the file; it uses them to build and export, more specifically. The Access bug in question refuses to drop those connections (even if you use .close). Access will either use up the connections and lock up, or if you're lucky will tell you no more connections and you can break it. 

For why this moves along if you delete "offending" files, each file is another connection. If you remove some, then there are more connections to use later, so it appears as if it's progressing further. Trust me, I about ripped my hair out before I discovered this on my own app, because previously working items just ... broke, so I started adding in code / tweaking things in a code panic trying to get it to stop, all to discover it's because connections (in this case, connection is local, and just means "connected to a memory slot" effectively, not a network connection, though that is also covered in this bug). 

As to the build log and incremental saving: ironically, having a build log that exports as it's written would make this problem worse, because it would connect to the outside file for each log entry. That said, I do think being able to save the log incrementally would be a good feature, and one I've occasionally wanted while troubleshooting. 

You can also turn on debug mode in the (Options- General Tab)[https://github.com/joyfullservice/msaccess-vcs-integration/wiki/Options#general-tab], which will provide the "stop when it fails" behavior you're talking about (I think).

**Comment by hecon5** [comment]
As a follow-up: Can you provide a copy of the "offending" file and a copy of the file that was able to be built so we could check that, too? While I'm reasonably certain it's a connection-caused bug, if there's something else, we should check that, too.

**Comment by alanv73** [comment]
I think I got this right... The file good_tbldef.zip is definitely the ones that I've just exported and are currently working fine. The file bad_tbldef.zip is downloaded from my github repo rolled back to Feb 2.

[good_tbldefs.zip](https://github.com/joyfullservice/msaccess-vcs-integration/files/8018860/good_tbldefs.zip)
[bad_tbldefs.zip](https://github.com/joyfullservice/msaccess-vcs-integration/files/8018863/bad_tbldefs.zip)
.

**Comment by hecon5** [comment]
OK, so ... this is odd.
The "Bad tables" are all HTML with a different extension. They load a GitHub page with the source; did you perchance save the page source, or check out the prior commit on your local repo to generate these? Once you load the HTML page, and copy out the displayed source, Comparing the source, they're identical, except some of them change case in the `Connect` String. that's it. 

I can't find any difference in the contents in the 3 I compared.

**Comment by alanv73** [comment]
Sorry, I think I've made that mistake before. I'll get the correct files
tomorrow if you think it will help.

Respectfully,

Alan Van Art

**Comment by hecon5** [comment]
We can give it a go, but if the line endings are the same, I'm suspecting your rebuild got lucky with the Access bug; I've also found some of the builds work, some don't, and when using [IslaDog's Available Connections](https://www.isladogs.co.uk/available-connections/index.html) tool, it appears (some) connections are released, but most are not. It appears that Access "eats" up a connection on its own at a rate of around 2-3 per 10 seconds (out of an available 255...what a small number), and the connections used by MSAccessVCS-Addin are (sometimes) released, it's that if the build takes longer than expected (>30 seconds in my case), Access "eats" the connections, so the build hangs.

**Comment by alanv73** [comment]
[bad_tbldefs2.zip](https://github.com/joyfullservice/msaccess-vcs-integration/files/8025782/bad_tbldefs2.zip)

**Comment by hecon5** [comment]
Same; only difference in the file is the case change from `common` to `Common`. Line endings and everything are identical.

**Comment by hecon5** [comment]
Do you have access to the file `\\arpmserv1\Common\Databases\TEST Inspections Database_be.accdb`? I'm wondering if the connection attempt failed (silently) due to out of connection issues (or anything, really), but carried on. 

If you have an intermittent connection to that file (network/vpn/etc.), would explain why build crashes.

As to the `_` character in table names, I have found that sometimes creates issues, as `_` (may) occasionally be interpreted as a class internal function, but during build, that shouldn't be an issue.

**Comment by alanv73** [comment]
Pretty sure I just figured out what's going on. At some point in the past I accidentally exported my DB with my local tables set to the name of the normally linked tables. This created an XML file for those tables along with the JSON and SQL files. Then when I corrected it later, renaming my local tables to _local, the XML files didn't get removed from the repo. So for these three tables links there are three source files in the repo, .SQL, .JSON, and .XML. If I remove the .XML files from the repo before building everything builds without issue even if I turn off 'Show Detailed Output'.

**Comment by hecon5** [comment]
@joyfullservice, do you think this is related to #296? @alanv73 can you update to the latest version and see if you can reproduce this issue? From what you said, I think there's a relation with it. Sorry about the confusion ...

**Comment by alanv73** [comment]
Yes, it sounds like the same issue. However, I found it odd that I did not end up with an orphaned JSON file from it. VCS must have removed the unneeded JSON file but not the unneeded XML file.

Not sure when I'll update, last time I did I couldn't build any of my repos exported with the older version.

**Comment by hecon5** [comment]
Understood; that was probably V3.4.15, which probably should've been a 4.0 release being it was a breaking change from the prior release. In any case, the latest should not affect building, but does address this issue.

**Comment by hecon5** [comment]
@joyfullservice: looks like Microsoft patched Access, so if @alanv73 is good with the fix, I think we can close this now.

**Comment by hecon5** [comment]
@joyfullservice: bump. This good to close?

**Comment by ad26495** [comment]
(for info; I know the issue is closed)

I am happy I found this issue, because I recently had exactly the same problems as @alanv73 (thank you!):
Failure to build during the tables phase (out of stack). Building OK when I selected ShowDetailedOutput. This behaviour was reproducing.
Based on the comments in this issue, I realized that I recently moved a table from local to linked but used the same name.
And indeed, there was an orphan .XML in the repo. Removed the orphan .XML and all is fine again: can build with ShowDetailedOutput un-selected.
